cupy/core/core.pyx

```diff
@@ -1168,16 +1168,142 @@ cdef class ndarray:
         return v
 
     def __setitem__(self, slices, value):
-        cdef ndarray v, x, y
+        """x.__setitem__(slices, y) <==> x[slices] = y
+
+        Supported ``slices`` consists of
+
+        * basic indexing
+
+        * one integer array
+
+        * combination of basic indexing and an integer array
+
+        .. note::
+
+            CuPy handles out-of-bounds indices differently from NumPy when
+            using integer array indexing.
+            NumPy handles them by raising an error, but CuPy wraps around them.
+
+            >>> import cupy
+            >>> x = cupy.arange(3)
+            >>> x[[1, 3]] = 10
+            >>> x
+            array([10, 10, 2])
+
+        .. note::
+
+            The behavior differs from NumPy when integer arrays in ``slices``
+            index same location multiple times.
+            In that case, the value that is actually stored is undefined.
+
+            >>> import cupy; import numpy
+            >>> a = cupy.zeros((2,))
+            >>> i = cupy.arange(10000) % 2
+            >>> v = cupy.arange(10000).astype(numpy.float)
+            >>> a[i] = v
+            >>> a
+            array([9150., 9151.])
+
+            On the other hand, NumPy stores the value corresponding to the
+            last index among the indices with duplicate target.
+
+            >>> import numpy
+            >>> a_cpu = numpy.zeros((2,))
+            >>> i_cpu = numpy.arange(10000) % 2
+            >>> v_cpu = numpy.arange(10000).astype(numpy.float)
+            >>> a_cpu[i_cpu] = v_cpu
+            >>> a_cpu
+            array([9998., 9999.])
+
+        """
+        cdef Py_ssize_t i, ndim, n_newaxes, n_ellipses, ellipsis, axis
+        cdef Py_ssize_t ellipsis_size
+        cdef ndarray a, v, x, y
+
         if not isinstance(slices, tuple):
-            slices = (slices,)
+            slices = [slices]
+        else:
+            slices = list(slices)
+
+        # Expand ellipsis into empty slices
+        ellipsis = -1
+        n_newaxes = n_ellipses = 0
+        for i, s in enumerate(slices):
+            if s is None:
+                n_newaxes += 1
+            elif s is Ellipsis:
+                n_ellipses += 1
+                ellipsis = i
+        ndim = self._shape.size()
+        noneslices = [slice(None)]
+        if n_ellipses > 0:
+            if n_ellipses > 1:
+                raise ValueError('Only one Ellipsis is allowed in index')
+            ellipsis_size = ndim - (<Py_ssize_t>len(slices) - n_newaxes - 1)
+            slices[ellipsis:ellipsis + 1] = noneslices * ellipsis_size
+
+        slices += noneslices * (ndim - <Py_ssize_t>len(slices) + n_newaxes)
+
+        if len(slices) > self.ndim + n_newaxes:
+            raise IndexError('too many indices for array')
+
+        # Check if advanced is true, and convert list/NumPy arrays to ndarray
```

Comment:
Is "Check if advanced is true, and convert list/NumPy arrays to cupy.ndarray" OK?
Since CuPy array is `cupy.ndarray`, I think that the sentence above is correct.